packages/block-library/src/image/index.php

```diff
@@ -0,0 +1,357 @@
+<?php
+/**
+ * core/image block.
+ */
+
+/**
+ * Warm the object cache with post and meta information for all found
+ * image blocks to avoid making individual database calls
+ * (similarly to `wp_make_content_images_responsive()`).
+ *
+ * @access private
+ * @since 4.2
+ *
+ * @param string $content The post content.
+ * @return string $content Unchanged post content.
+ */
+function _gutenberg_cache_images_meta( $content ) {
+	// Need to find all image blocks and get the attachment IDs from them BEFORE the parser is run
+	// so we can get the image attachments meta all at once from the DB.
+	if ( preg_match_all( '/^<!-- wp:image {.*"id":(\d+),.*} -->$/m', $content, $matches ) ) {
```

Comment:
Is this something where being able to hook to the parse (#10108) would prevent us from needing to try to match via a regular expression?

Reply:
To cache the attachments meta we will need to know all attachments IDs before the meta is needed. The point is to do one call to the DB for all the meta data, not multiple calls. We would need to process all "image" blocks, all "media and text" (when an image is used), "gallery", and "cover".

I know it's somewhat hacky to use a regex, but not sure we can get all these IDs before callbacks have started/blocks are processed.

Reply:
That seems precisely the intent of the opportunity provided through `block_pre_render` proposed in #10108.

Reply:
Maybe then in addition to `block_pre_render` it would be useful to have a filter which occurs immediately following the `do_blocks` parse, before the loop to render each individual block.

cc @dmsnell

Reply:
> fileWidth, fileHeight, srcSet, sizes are redundant if the id is already known.

Well, you'd think so but images (and image meta) is "the land of thousand edge cases" :) Example: the website was "exported" and then imported into another site. Now all attachment IDs stored in post_content are "wrong". (Well in this case WP will also fail to generate the srcset).

It would be beneficial to have some more info/meta about the image from the moment it was added to the post. Also, that somewhat helps with external images.

The `srcSet` and `sizes` would be pretty nice to have inside the editor but not essential for now. I'm almost ready with the next chunk split off this PR, there they are not used. However it has `fileWidth`, `fileHeight`, `userWidth`, `userHeight` for when the user set these manually (can reuse width and height and just add one prop indicating this), and `editWidth` that lets the theme scale images according to the width of the block in the editor when the post was created. Of course, none of these are set in stone, lets move the discussion to that pr (I'll ping you there).

Reply:
> in addition to block_pre_render it would be useful to have a filter which occurs immediately following the do_blocks parse, before the loop to render each individual block.

this is how I imagined we would do things like what's being discussed here. if we want to do post-wide operations or side-effecting transformations then we can filter each block and hold state in the filter function with a static variable or class property, then a closing operation at the end. I hadn't thought about placing that last filter though in this context, so thanks for pointing it out - I'll add it to #10108